dwave/optimization/_model.pyx

```diff
@@ -251,6 +233,37 @@ cdef class _Graph:
 
         return model
 
+    @staticmethod
+    def _from_file_header(file):
+        """Read/validate the header from a model file.
+
+        Advances the stream position to the end of the header.
+
+        Returns:
+            The serialization version as a 2-tuple.
+
+            The header data as a dict.
+        """
+        prefix = b"DWNL"
+
+        read_prefix = file.read(len(prefix))
+        if read_prefix != prefix:
+            raise ValueError("Unknown file type. Expected magic string "
+                             f"{prefix!r} but received {read_prefix!r} "
+                             "instead")
+
+        version = tuple(file.read(2))
+
+        if version not in KNOWN_SERIALIZATION_VERSIONS:
+            raise ValueError("Unknown serialization format. Expected one of "
+                             f"{KNOWN_SERIALIZATION_VERSIONS} but received {version} "
+                             "instead. Upgrading your dwave-optimization version may help.")
+
+        header_len = struct.unpack('<I', file.read(4))[0]
+        header_data = json.loads(file.read(header_len).decode('ascii'))
```

Comment:
Perhaps catch JSON decoding errors / insufficient data, wrapping them with some kind of `InvalidFileFormat` exception?

Rationale: this is a helper function that returns (version, data) from a file, so handling errors locally along the way would be nice.